### ISSUE TITLE
hierarchical multilabel classification head

### ISSUE THREAD

**Issue by cregouby** [mlverse/tabnet#121]
Hierarchical Multilabel Classification (HMC) provides improvement over naive classification in a lot of cases.

This requires to Implement the HMC head and loss function like in [C-HMCNN](https://github.com/EGiunchiglia/C-HMCNN)

This includes
- [x] consider the best hierarchical data format for ~~outcome~~ dataset (like [`data.tree`](https://cran.r-project.org/web/packages/data.tree/vignettes/data.tree.html#introduction) ~~or [`tidygraph`](https://tidygraph.data-imaginist.com/)~~ )
- [x] choose an example dataset
- [x] create a `tabnet_fit.Node()` method
- [x] allow for `valid_split =` in tabnet_fit.Node()
- [x] fix a `tabnet_pretrain.Node()` void method
- [x] create a `predict.tabnet()` method for Node
- [x] create a `tabnet_explain()` method for Node()
- [x] create a predict method for data = Node
- [x] implement the R matrix computing from data.tree
- [x] implement the MCM loss
- [ ] create an inspiring vignette

**Comment by phgrosjean** [comment]
@cregouby As we discussed in Avignon, I can propose a dataset for playing with hierarchical labels. This is zooplankton classification as manipulated by the {zooimage} package available on CRAN. A short example training dataset can be downloaded from: https://filedn.com/lzGVgfOGxb6mHFQcRn9ueUb/zooimage/zooplankton_train.rds.

- labels are in `Class`
- attributes are the other columns
- the hierarchy of labels are in attr(<dataset>, "path")`. For instance, there is "Copepoda/Podoplea/Poecilostomatoida" which means that `Class` "Poecilostomatoida" in the data.frame is part of a supercategory "Podoplea" (together with "Cyclopoida" and "Harpacticoida"). At an even higher level, "Copepoda" contains both "Podoplea" and "Gymnoplea". This is a very basic way of encoding the hierarchy of labels. Obviously this needs to be transformed into a more useable object for {tabnet}.

This small training set is featured in the book Data Mining Applications with R, Chap.12 (see: https://shop.elsevier.com/books/data-mining-applications-with-r/zhao/978-0-12-411511-8). The full data with images of each individual plankton item and two more test sets is available from http://www.sciviews.org/zooimage/Data%20mining%20with%20R/Data%20Mining%20with%20R.zip, together with R code to play with these as explained in the book.

The training set may be too small for {tabnet}. We have much larger datasets available (up to 1.5 millions rows) if needed. This is to get an idea of the kind of data I can propose.

**Comment by cregouby** [comment]
Hello @phgrosjean,

I'm very glad we can continue our discussion from the RR_2023 and start working around a real dataset ! 
I'll use it for the vignette if it's OK for you (because the current starwars dataset is giving me headache... )

**Comment by phgrosjean** [comment]
@cregouby Yes, yes, use it! You can cite the book as a reference.
